### ISSUE TITLE
Disallow creating an invalid Ident

### ISSUE THREAD

**Issue by dtolnay** [dtolnay/syn#141]
An instance of Ident should always contain a valid identifier.

**Comment by mystor** [comment]
Should we reject keywords in `Ident`s like `as` or `fn`? I assume not as they are useful to keep in `Ident`s sometimes.

**Comment by dtolnay** [comment]
I suspect that Ident isn't going to exist anymore, we will just use the Term type from https://github.com/rust-lang/rust/pull/40939 directly. That one can contain keywords.

**Comment by mystor** [comment]
Currently on master `Ident` is a distinct type which bundles up `Symbol` (or `Term` if the name gets changed) and a `Span`. We'll probably keep `Ident` for that, but will probably change `Lifetime` to be a `(Symbol, Span)` instead of an `Ident`.

**Comment by dtolnay** [comment]
Ah you're right, of course it needs a Span. I would allow anything that can match `$ident` in macro_rules, which seems to include keywords.

```rust
macro_rules! ident {
    ($i:ident) => {}
}

fn main() {
    ident!(mystor);
    ident!(fn);
}
```

**Comment by dtolnay** [comment]
Fixed in 570695ea9345d9db9e089a9e21f2b157dffa6fb3.
